### PR TITLE
Watching versions of `bevy/get` and `bevy/list` with HTTP SSE

### PR DESCRIPTION
**Pull request by LiamGallagher737**
## Objective

Add a way to stream BRP requests when the data changes.

#### BRP Side (reusable for other transports)

Add a new method handler type that returns a optional value. This handler is run in update and if a value is returned it will be sent on the message channel. Custom watching handlers can be added with `RemotePlugin::with_watching_method`.

#### HTTP Side

If a request comes in with `+watch` in the method, it will respond with `text/event-stream` rather than  a single response.

## Testing

I tested with the podman HTTP client. This client has good support for SSE's if you want to test it too.

## Parts I want some opinions on

- For separating watching methods I chose to add a `+watch` suffix to the end kind of like `content-type` headers. A get would be `bevy/get+watch`.
- Should watching methods send an initial response with everything or only respond when a change happens? Currently the later is what happens.

## Future work

- The `bevy/query` method would also benefit from this but that condition will be quite complex so I will leave that to later.

### CODE REVIEW

**Comment by JMS55** [comment]
Really not sure how I feel about SSE vs a proper streaming protocal like QUIC or WebTransport...

**Comment by LiamGallagher737** [comment]
> Really not sure how I feel about SSE vs a proper streaming protocol like QUIC or WebTransport...

I chose it just due to being easy to implement with the existing HTTP transport. I am happy to swap it for something else as it's the BRP part I care about, the SSE is just to have something to show it works.

**Comment by LiamGallagher737** [comment]
After a quick look at the [wtransport](https://github.com/BiagioFesta/wtransport) crate, it is defiantly something I'd like to have a go at integrating. Until then I think the SSE serves as a good example of how to use the streaming.

**Comment by LiamGallagher737** [comment]
@notmd I'd also appreciate a review from you since you have been working on streaming stuff.

**Comment by notmd** [comment]
I'm not sure about mixing streaming and one-time request into 1 handler is a good I idea. Also, letting the transport control weather to stream seems bad. The `RemoteMethods` should act as contracts for what a method does, either streaming or one-time. And the transport should return err if it not support streaming.

**Comment by LiamGallagher737** [comment]
Currently re-working it to separate the streaming handlers and only stream changed data, not entire responses.

**Comment by LiamGallagher737** [comment]
One issue with this is some components such as `bevy_render::view::visibility::ViewVisibility` are "changed" every frame and a response is sent even though the value of `true` stays the same. I am guessing this is because it is being accessed mutably even though the value isn't changed. Does anyone know a solution to this?

**Comment by djeedai** [comment]
If a component gets flagged by the change detection system without an actual change, this is a bug in that component management, and likely breaks other uses of it. I'd log a bug. Otherwise we can't rely on change detection.

**Review by djeedai — APPROVED** [review]

**Comment by LiamGallagher737** [comment]
As a note on bandwidth usage, requesting `bevy/get+watch` with `Transform` as the only component on the cube result in roughly 6KB per second with the app running at 60fps.

File: examples/remote/server.rs
Comment (LiamGallagher737) [reply examples/remote/server.rs]: I made the movement continuous. 

As for what to do if the volume of data gets to much for the transport, no clue. Currently due to using `force_send` on the senders I believe un-received messages are overridden by shifting the queue down (removing the oldest one) and adding the new message to the end. Not the best way of handling it but it should work for now until someone thinks of something better.

**Comment by notmd** [comment]
> One issue with this is some components such as `bevy_render::view::visibility::ViewVisibility` are "changed" every frame and a response is sent even though the value of `true` stays the same. I am guessing this is because it is being accessed mutably even though the value isn't changed. Does anyone know a solution to this?

I have the same issue in my impl. There is `track_change_detection` feature, but I haven't dig into yet. Searching `mut ViewVisibility` in the repo have few systems use it, probably one of them

**Comment by pablo-lua** [comment]
> One issue with this is some components such as `bevy_render::view::visibility::ViewVisibility` are "changed" every frame and a response is sent even though the value of `true` stays the same. I am guessing this is because it is being accessed mutably even though the value isn't changed. Does anyone know a solution to this?

ViewVisibility is a Component used for calculations and therefore, is set every frame to false, and after that, other components do they thing and will se this to True after. 
Would probably be better to bypass this when setting, but I'm not sure if this is the best approach. For that kind of component, I think the client side should probably be responsible for storing an cache and seeing if the value was in fact changed. We could do this kind of cache in the backend, but I think this would be much more tricky for this PR

**Comment by LiamGallagher737** [comment]
Found an issue with missing removal events so I'll rework that. Please don't merge until then.

**Comment by LiamGallagher737** [comment]
Good to go now

**Review by bushrat011899 — CHANGES_REQUESTED** [review]
Don't have time just yet to do a test locally, but I have some suggestions already mostly around avoid performance regression.

File: crates/bevy_remote/src/lib.rs
Comment (LiamGallagher737) [reply crates/bevy_remote/src/lib.rs:829]: Good to know! I'll make this PR once this one is merged

**Review by bushrat011899 — APPROVED** [review]
Can confirm this works as expected. Sending `bevy/query` to find the cube `Entity` and then using:

```json
{
    "jsonrpc": "2.0",
    "method": "bevy/get+watch",
    "params": {
        "entity": 4294967333,
        "components": [
            "bevy_transform::components::transform::Transform"
        ]
    }
}
```

does start an HTTP stream sending results continuously (since in the `server` example the cube's transform is changing continuously). I think this is a solid piece of infrastructure to include, even if it isn't based on something more feature rich like web sockets.

**Pull request merged**
